logger.go

```diff
@@ -74,52 +74,54 @@ func NewLogger(ctx context.Context) *Logger {
 	return l
 }
 
-// ServeHTTP calls the next function and incorporates the response into its response cache.
+// Middleware returns a handler that incorporates the response into its response cache.
 // If next panics the panic is recovered and logged.
-func (l *Logger) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
-	start := time.Now()
-	reqID := <-l.ids
-	r = setCtxRequestId(reqID, r)
-	r = setStartAtTime(r, start)
-
-	remote := r.Header.Get(remoteAddrHeaderName)
-	if remote == "" {
-		remote = r.RemoteAddr
-	}
+func (l *Logger) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		reqID := <-l.ids
+		r = setCtxRequestId(reqID, r)
+		r = setStartAtTime(r, start)
+
+		remote := r.Header.Get(remoteAddrHeaderName)
+		if remote == "" {
+			remote = r.RemoteAddr
+		}
 
-	defer func() {
-		if err := recover(); err != nil {
-			if rw.Header().Get("Content-Type") == "" {
-				rw.Header().Set("Content-Type", "text/plain; charset=utf-8")
+		defer func() {
+			if err := recover(); err != nil {
+				if rw.Header().Get("Content-Type") == "" {
+					rw.Header().Set("Content-Type", "text/plain; charset=utf-8")
+				}
+
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				grip.Critical(message.Fields{
+					"stack":    message.NewStack(2, "").Raw(),
+					"panic":    err,
+					"action":   "aborted",
+					"request":  reqID,
+					"duration": time.Since(start),
+					"span":     time.Since(start).String(),
+					"remote":   remote,
+					"path":     r.URL.Path,
+				})
 			}
-
-			rw.WriteHeader(http.StatusInternalServerError)
-
-			grip.Critical(message.Fields{
-				"stack":    message.NewStack(2, "").Raw(),
-				"panic":    err,
-				"action":   "aborted",
-				"request":  reqID,
-				"duration": time.Since(start),
-				"span":     time.Since(start).String(),
-				"remote":   remote,
-				"path":     r.URL.Path,
-			})
-		}
-	}()
-
-	next(rw, r)
-
-	grip.ErrorWhen(
-		sometimes.Percent(logErrorPercentage),
-		message.WrapError(
-			l.addToCache(rw, r),
-			message.Fields{
-				"message": "adding response to buffer",
-				"path":    r.URL.Path,
-			},
-		),
-	)
+		}()
+
+		next.ServeHTTP(rw, r)
+
+		grip.ErrorWhen(
+			sometimes.Percent(logErrorPercentage),
+			message.WrapError(
+				l.addToCache(rw, r),
```

Comment:
This ends up in Splunk, right?

Reply:
Specifically errors returned from addToCache do?

Reply:
Yes, 10% of the errors from addToCache will be logged to Splunk. So like if we're filling the buffer faster than we can process it we'll see [response buffer is full](https://github.com/evergreen-ci/logkeeper/blob/4c355d457549c3a3ff8b06eb84ed347b38f6ec99/logger.go#L183) messages [there](https://mongodb.splunkcloud.com/en-US/app/search/search?q=search%20index%3Dlogkeeper%20metadata.level%20%3E%3D%2070%20message%3D%22adding%20response%20to%20buffer%22&sid=1658955422.3469948&display.page.search.mode=verbose&dispatch.sample_ratio=1&workload_pool=standard_perf&earliest=-30m%40m&latest=now).